### PR TITLE
generic notebook for 3D basemap solution, for any object of interest.

### PR DESCRIPTION
**Pull request by vbhv14**
[temp PR]
Story:

A small fraction of users will opt for this notebook workflow, in the solution.
Who are not getting satisfactory results with the shipped production model.

Points considered for this notebook:

- [x] Sample notebook does not go into details about `predict_las( )`, as there will be an inferencing tool present in the solutions.
**Context:** 
It will reduce the importance of toolbox.  _[covered in guide & sample, for which the link is provided, in this sample]_

- [x] Sample does not go into details about best practices & in-depth explanation of workflow,  as the target user wants to train a production-grade model, that can be used in the solution. Which is already provided with the solution.
**Context:** 
It will reduce the importance of the shipped production model, with the solution. _[covered in guide & sample, for which the link is provided, in this sample]_

- [x] The Sample is intended to work or used by the user in ArcGIS Pro, and `show_batch( ) `& `show_results( )` don't work in PRO, hence that part is also not included.
**Context:** 
show_batch( ) & show_results( ) don't work.

### CODE REVIEW

**Comment by priyankatuteja** [comment]
@BP-Ent Could you please help in reviewing this notebook?

**Comment by priyankatuteja** [comment]
@BP-Ent Could you please help in reviewing this notebook?

File: samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb [hunk 1]
```diff
@@ -0,0 +1,383 @@
+{
```
Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <h1>Training a Deep Learning model for Point Cloud Classification</h1><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>One of the most important initial steps in the workflow of '3D Basemap' solutions, is the classification of LiDAR Point Clouds. This sample notebook provides a reference for how to train a deep learning model for point cloud classification. </p><p> </p><p>To achieve this, the ArcGIS API for Python's implementation of PointCNN is used. Further details on the PointCNN implementation in the API (working principle, architecture, best practices, etc.), can be found <a href="https://developers.arcgis.com/python/guide/point-cloud-segmentation-using-pointcnn" target="_blank">here</a>, along with instructions on how to set up the Python environment. Additional sample notebooks related to PointCNN can be found <a href="https://developers.arcgis.com/python/sample-notebooks/" target="_blank">here</a>.</p><p> </p><p>Before proceeding through this notebook, it is advised to go through the API Reference for PointCNN [<code>export_point_dataset( ), prepare_data( ) Transform3d( ), &amp; PointCNN( )</code>], that can be found <a href="http://developers.arcgis.com/python/api-reference/arcgis.learn.html" target="_blank">here</a>. It will help in understanding the PointCNN's workflow in detail. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <ol><li>Convert the dataset into&nbsp;<code>.las</code>&nbsp;format.<em>&nbsp;(Refer to ArcGIS Pro's&nbsp;</em><a href="https://pro.arcgis.com/en/pro-app/tool-reference/conversion/convert-las.htm" target="_blank"><em>'Convert LAS'</em></a><em>&nbsp;GP tool.)</em></li><li>Split all the&nbsp;<code>.las</code>&nbsp;files into three unique sets, one for training, one for validation, and one for testing. There is no fixed rule, but generally, the validation data for point cloud in&nbsp;<code>.las</code>&nbsp;format should be at least 1 % (by size) of the total data available, with appropriate diversity within the validation dataset.<em>&nbsp;(For ease in splitting the big&nbsp;</em><code><em>.las</em></code><em>&nbsp;files into the appropriate ratios, ArcGIS Pro's&nbsp;</em><a href="https://pro.arcgis.com/en/pro-app/tool-reference/3d-analyst/tile-las.htm" target="_blank"><em>'Tile LAS'</em></a><em>&nbsp;GP tool can be used.)</em></li><li>Reassign classcodes if needed. Consider the 'object of interest' or any requirement related to the workflow of the 3D Basemap solution.<em>&nbsp;(Refer to ArcGIS Pro's&nbsp;</em><a href="https://pro.arcgis.com/en/pro-app/tool-reference/3d-analyst/change-las-class-codes.htm" target="_blank"><em>'Change LAS Class Codes'</em></a><em>&nbsp;GP tool.)</em></li></ol><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>In this step, <code>.las</code> files are converted to 'HDF5 binary data format'. The function export_point_dataset( )is used for this. </p><p> </p><p>Apart from that, the folder structure for the <code>data_path</code> argument must contain two folders, 'train' &amp; 'val', which should have the training and validation files, respectively. <em>(as shown in figure 1)</em></p><p> </p><p>The <code>export_point_dataset( )</code> function only needs an 'input folder' path and an 'output folder' path. The rest of the arguments have an appropriate default value, which should be valid in most cases. Here, 'extra_features' is used to include additional attributes from LiDAR data, like intensity, number of returns, etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>This section is divided into two parts, preparing the training data and training the model.</p><p> </p><p>For <code>prepare_data()</code>, the argument 'batch_size' controls the number of training samples to be processed at one time. If you have a powerful GPU, this number can be further increased.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>The <code>lr_find()</code>function is used to control the rate at which existing information will be overwritten by newly acquired information throughout the training process. If no value is specified, the optimal learning rate will be extracted from the learning curve during the training process.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>In <code>fit( )</code>, the argument 'epochs' controls the maximum number of epochs for which the model will be trained. A maximum epoch of one means the dataset will be passed forward and backward through the neural network one time. <s>And</s> If 'early_stopping' is True, then the model training will stop when the model is no longer improving, regardless of the 'epochs' parameter value specified.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>The <code>save()</code> function is used to the save the trained model. The saved folder will contain the ESRI model definition file, which will be required by the "Deep Learning Tools" in the 3D Basemap Solution as an input for referencing. Alternatively, <code>predict_las( )</code> can be used for inferencing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: 'corrected accordingly'', Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: 'corrected accordingly'', Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly' Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

**Comment by vbhv14** [comment]
thanks, @BP-Ent for the review, I have made the changes, you mentioned.
and added a final paragraph to complete the story. 
kindly, let me know about any further changes, which are left.

**Comment by vbhv14** [comment]
cc @priyankatuteja @rohitgeo  
With Reference to https://github.com/ArcGIS/geosaurus/issues/3368

In the final released version of Pro 2.6, we have issues related to PointCNN's `show_batch()` & `show_results()` & fastprogress. (DL installer was used)

Here are the relevant screenshots related to existing issues.
showbtch_res_error
fast progress error
fast progress error 2

While `lr_find()` works with a popup.

lr find error

File: samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
```
Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: <p>Once the model has been saved, it is then ready to be implemented in the deep learning tools provided with the solution. As the deep learning tools are object-specific, it is important to use the appropriate tool. For example, if a user trains a model specifically for tree points, then the 'classify tree points using deep learning' tool should be used. The provided object-specific deep learning tools also expect the model's classcode to be specific. For example, the 'classify tree points using deep learning' tool expects the user-trained model's classcode for tree points to be '5'. The user-trained model's classcode can be remapped to some other classcode when writing in the <code>.las</code> files.</p><p> </p><p>This notebook has summarized the workflow for the training of a deep learning model for point cloud classification, intended to be used when the bundled models in the 3D Basemap solution are not utilized.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/763/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

**Comment by vbhv14** [comment]
@BP-Ent , thanks, pushed the final changes.

File: samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb (hunk 2 above)
Comment (vbhv14) [reply samples/04_gis_analysts_data_scientists/Training a Deep Learning model, for using it in 3D basemap Solutions..ipynb:1]: corrected accordingly, Thanks.

**Comment by vbhv14** [comment]
closing the PR,
the deliverables are handed over.
thanks, everyone for helping with the reviews.
